File: app/src/main/java/com/thewizrd/simpleweather/notifications/WeatherNotificationBuilder.java
package com.thewizrd.simpleweather.notifications;

import android.app.Notification;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.text.TextUtils;
import android.view.View;
import android.widget.RemoteViews;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;

import com.thewizrd.shared_resources.controls.DetailItemViewModel;
import com.thewizrd.shared_resources.controls.WeatherDetailsType;
import com.thewizrd.shared_resources.controls.WeatherNowViewModel;
import com.thewizrd.shared_resources.icons.WeatherIconProvider;
import com.thewizrd.shared_resources.icons.WeatherIcons;
import com.thewizrd.shared_resources.icons.WeatherIconsManager;
import com.thewizrd.shared_resources.icons.WeatherIconsProvider;
import com.thewizrd.shared_resources.utils.Colors;
import com.thewizrd.shared_resources.utils.ConversionMethods;
import com.thewizrd.shared_resources.utils.ImageUtils;
import com.thewizrd.shared_resources.utils.NumberUtils;
import com.thewizrd.shared_resources.utils.SettingsManager;
import com.thewizrd.shared_resources.utils.StringUtils;
import com.thewizrd.shared_resources.utils.Units;
import com.thewizrd.shared_resources.utils.WeatherUtils;
import com.thewizrd.simpleweather.App;
import com.thewizrd.simpleweather.R;
import com.thewizrd.simpleweather.main.MainActivity;

public class WeatherNotificationBuilder {
    private static final String TAG = "WeatherNotificationBuilder";

    static Notification updateNotification(String notChannelID, @NonNull final WeatherNowViewModel viewModel) {
        final Context context = App.getInstance().getAppContext();
        final SettingsManager settingsManager = App.getInstance().getSettingsManager();
        final WeatherIconsManager wim = WeatherIconsManager.getInstance();

        // Build update
        RemoteViews updateViews = new RemoteViews(context.getPackageName(), R.layout.weather_notification_layout);

        String condition = viewModel.getCurCondition();
        String hiTemp = StringUtils.removeNonDigitChars(viewModel.getHiTemp());
        String loTemp = StringUtils.removeNonDigitChars(viewModel.getLoTemp());
        String temp = viewModel.getCurTemp() != null ?
                StringUtils.removeNonDigitChars(viewModel.getCurTemp()) : WeatherIcons.PLACEHOLDER;

        // Weather icon
        updateViews.setImageViewResource(R.id.weather_icon, wim.getWeatherIconResource(viewModel.getWeatherIcon()));

        // Location Name
        updateViews.setTextViewText(R.id.location_name, viewModel.getLocation());

        // Condition text
        updateViews.setTextViewText(R.id.condition_weather,
                String.format("%s°%s - %s", !StringUtils.isNullOrWhitespace(temp) ? temp : WeatherIcons.PLACEHOLDER, viewModel.getTempUnit(), condition));

        // Details
        updateViews.setTextViewText(R.id.condition_hi, StringUtils.containsDigits(hiTemp) ? (hiTemp + "°") : WeatherIcons.PLACEHOLDER);
        updateViews.setTextViewText(R.id.condition_lo, StringUtils.containsDigits(loTemp) ? (loTemp + "°") : WeatherIcons.PLACEHOLDER);
        updateViews.setViewVisibility(R.id.condition_hilo_layout, viewModel.isShowHiLo() ? View.VISIBLE : View.GONE);

        // Get extras
        DetailItemViewModel chanceModel = null, windModel = null, feelsLikeModel = null, humidityModel = null, popRainModel = null, popSnowModel = null;
        for (DetailItemViewModel input : viewModel.getWeatherDetails()) {
            if (input.getDetailsType() == WeatherDetailsType.POPCHANCE) {
                chanceModel = input;
            } else if (chanceModel == null && input.getDetailsType() == WeatherDetailsType.POPCLOUDINESS) {
                chanceModel = input;
            } else if (input.getDetailsType() == WeatherDetailsType.WINDSPEED) {
                windModel = input;
            } else if (input.getDetailsType() == WeatherDetailsType.FEELSLIKE) {
                feelsLikeModel = input;
            } else if (input.getDetailsType() == WeatherDetailsType.HUMIDITY) {
                humidityModel = input;
            } else if (input.getDetailsType() == WeatherDetailsType.POPRAIN) {
                popRainModel = input;
            } else if (input.getDetailsType() == WeatherDetailsType.POPSNOW) {
                popSnowModel = input;
            }

            if (chanceModel != null && windModel != null && feelsLikeModel != null && humidityModel != null && popRainModel != null && popSnowModel != null) {
                break;
            }
        }

        // Extras
        if (chanceModel != null) {
            updateViews.setImageViewResource(R.id.weather_popicon, chanceModel.getIcon());
            updateViews.setTextViewText(R.id.weather_pop, chanceModel.getValue());
            updateViews.setViewVisibility(R.id.weather_pop_layout, View.VISIBLE);
        } else {
            updateViews.setViewVisibility(R.id.weather_pop_layout, View.GONE);
        }
        if (windModel != null) {
            final int windIconResId = wim.getWeatherIconResource(WeatherIcons.WIND_DIRECTION);
            if (windModel.getIconRotation() != 0) {
                updateViews.setImageViewBitmap(R.id.weather_windicon,
                        ImageUtils.rotateBitmap(ImageUtils.bitmapFromDrawable(context, windIconResId), windModel.getIconRotation())
                );
            } else {
                updateViews.setImageViewResource(R.id.weather_windicon, windIconResId);
            }
            String speed = TextUtils.isEmpty(windModel.getValue()) ? "" : windModel.getValue().toString();
            speed = speed.split(",")[0];
            updateViews.setTextViewText(R.id.weather_windspeed, speed);
            updateViews.setViewVisibility(R.id.weather_wind_layout, View.VISIBLE);
        } else {
            updateViews.setViewVisibility(R.id.weather_wind_layout, View.GONE);
        }
        updateViews.setViewVisibility(R.id.extra_layout, chanceModel != null || windModel != null ? View.VISIBLE : View.GONE);

        // Extras 2
        RemoteViews bigUpdateViews;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            bigUpdateViews = new RemoteViews(updateViews);
        } else {
            bigUpdateViews = updateViews.clone();
        }
        if (feelsLikeModel != null) {
            bigUpdateViews.setTextViewText(R.id.feelslike_label, context.getString(R.string.label_feelslike));
            bigUpdateViews.setTextViewText(R.id.feelslike_temp, feelsLikeModel.getValue() + viewModel.getTempUnit());
            bigUpdateViews.setViewVisibility(R.id.feelslike_layout, View.VISIBLE);
        }
        if (humidityModel != null) {
            bigUpdateViews.setImageViewResource(R.id.humidity_icon, humidityModel.getIcon());
            bigUpdateViews.setTextViewText(R.id.humidity, humidityModel.getValue());
            bigUpdateViews.setViewVisibility(R.id.humidity_layout, View.VISIBLE);
        }
        if (popRainModel != null) {
            bigUpdateViews.setImageViewResource(R.id.precip_rain_icon, popRainModel.getIcon());
            bigUpdateViews.setTextViewText(R.id.precip_rain, popRainModel.getValue());
            bigUpdateViews.setViewVisibility(R.id.precip_rain_layout, View.VISIBLE);
        }
        if (popSnowModel != null) {
            bigUpdateViews.setTextViewText(R.id.precip_snow, popSnowModel.getValue());
            bigUpdateViews.setViewVisibility(R.id.precip_snow_layout, View.VISIBLE);
        }
        bigUpdateViews.setViewVisibility(R.id.extra2_layout, feelsLikeModel != null || humidityModel != null || popRainModel != null || popSnowModel != null ? View.VISIBLE : View.GONE);

        NotificationCompat.Builder mBuilder =
                new NotificationCompat.Builder(context, notChannelID)
                        .setCustomContentView(updateViews)
                        .setCustomBigContentView(bigUpdateViews)
                        .setPriority(NotificationCompat.PRIORITY_LOW)
                        .setOngoing(true)
                        .setShowWhen(false);

        if (settingsManager.getNotificationIcon().equals(SettingsManager.TEMPERATURE_ICON)) {
            Integer tempLevel = NumberUtils.tryParseInt(temp.replace("°", ""));

            if (tempLevel == null) {
                mBuilder.setSmallIcon(R.drawable.notification_temp_unknown);
            } else {
                mBuilder.setSmallIcon(WeatherNotificationTemp.getTempDrawable(tempLevel));
            }
        } else if (settingsManager.getNotificationIcon().equals(SettingsManager.CONDITION_ICON)) {
            if (wim.isFontIcon()) {
                mBuilder.setSmallIcon(wim.getWeatherIconResource(viewModel.getWeatherIcon()));
            } else {
                // Use default icon pack here; animated icons are not supported here
                WeatherIconProvider wip = WeatherIconsManager.getProvider(WeatherIconsProvider.KEY);
                mBuilder.setSmallIcon(wip.getWeatherIconResource(viewModel.getWeatherIcon()));
            }
        }

        Intent onClickIntent = new Intent(context, MainActivity.class)
                .setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        PendingIntent clickPendingIntent = PendingIntent.getActivity(context, 0, onClickIntent, 0);
        mBuilder.setContentIntent(clickPendingIntent);

        // Builds the notification and issues it.
        Float temp_float = NumberUtils.tryParseFloat(temp);
        if (temp_float != null) {
            float temp_f = viewModel.getCurTemp().endsWith(Units.FAHRENHEIT) ?
                    temp_float : ConversionMethods.CtoF(temp_float);
            mBuilder.setColor(WeatherUtils.getColorFromTempF(temp_f));
        } else {
            mBuilder.setColor(Colors.SIMPLEBLUE);
        }
        return mBuilder.build();
    }
}

File: app/src/main/java/com/thewizrd/simpleweather/notifications/WeatherNotificationBuilder.kt
package com.thewizrd.simpleweather.notifications

import android.app.Notification
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.drawable.Icon
import android.os.Build
import android.text.SpannableStringBuilder
import android.text.TextUtils
import android.view.View
import android.widget.RemoteViews
import androidx.annotation.IdRes
import androidx.annotation.LayoutRes
import androidx.core.app.NotificationCompat
import com.thewizrd.common.controls.*
import com.thewizrd.common.utils.ImageUtils.bitmapFromDrawable
import com.thewizrd.common.utils.ImageUtils.rotateBitmap
import com.thewizrd.shared_resources.di.settingsManager
import com.thewizrd.shared_resources.icons.WeatherIcons
import com.thewizrd.shared_resources.icons.WeatherIconsEFProvider
import com.thewizrd.shared_resources.locationdata.LocationData
import com.thewizrd.shared_resources.preferences.SettingsManager
import com.thewizrd.shared_resources.sharedDeps
import com.thewizrd.shared_resources.utils.Colors
import com.thewizrd.shared_resources.utils.ConversionMethods
import com.thewizrd.shared_resources.utils.StringUtils.containsDigits
import com.thewizrd.shared_resources.utils.StringUtils.isNullOrWhitespace
import com.thewizrd.shared_resources.utils.StringUtils.removeNonDigitChars
import com.thewizrd.shared_resources.utils.Units
import com.thewizrd.shared_resources.utils.getColorFromTempF
import com.thewizrd.simpleweather.R
import com.thewizrd.simpleweather.main.MainActivity
import com.thewizrd.weather_api.weatherModule
import java.time.ZonedDateTime
import java.time.temporal.ChronoUnit
import kotlin.math.min

object WeatherNotificationBuilder {
    private const val TAG = "WeatherNotificationBuilder"
    private const val MAX_FORECASTS = 6

    suspend fun updateNotification(
        context: Context,
        notChannelID: String,
        location: LocationData,
        viewModel: WeatherUiModel
    ): Notification {
        val wim = sharedDeps.weatherIconsManager
        val wip = wim.getIconProvider(WeatherIconsEFProvider.KEY)

        // Build update
        val updateViews = RemoteViews(
            context.packageName, if (wim.isFontIcon) {
                R.layout.weather_notification_layout_tintable
            } else {
                R.layout.weather_notification_layout
            }
        )

        val condition = viewModel.curCondition
        val hiTemp = viewModel.hiTemp?.removeNonDigitChars()
        val loTemp = viewModel.loTemp?.removeNonDigitChars()
        val temp = viewModel.curTemp?.removeNonDigitChars() ?: WeatherIcons.PLACEHOLDER
        val weatherIconResId = wim.getWeatherIconResource(viewModel.weatherIcon)

        // Weather icon
        updateViews.setImageViewResource(R.id.weather_icon, weatherIconResId)

        // Location Name
        updateViews.setTextViewText(R.id.location_name, viewModel.location)

        // Condition text
        updateViews.setTextViewText(
            R.id.condition_weather,
            String.format(
                "%s° - %s",
                temp.ifBlank { WeatherIcons.PLACEHOLDER },
                condition
            )
        )

        // Details
        // Get extras
        val chanceModel = viewModel.weatherDetailsMap[WeatherDetailsType.POPCHANCE]
            ?: viewModel.weatherDetailsMap[WeatherDetailsType.POPCLOUDINESS]
        val windModel = viewModel.weatherDetailsMap[WeatherDetailsType.WINDSPEED]
        val windGustModel = viewModel.weatherDetailsMap[WeatherDetailsType.WINDGUST]
        val feelsLikeModel = viewModel.weatherDetailsMap[WeatherDetailsType.FEELSLIKE]
        val humidityModel = viewModel.weatherDetailsMap[WeatherDetailsType.HUMIDITY]

        val conditionTempBuilder = SpannableStringBuilder()

        if (viewModel.isShowHiLo) {
            conditionTempBuilder.append(
                String.format(
                    "%s / %s",
                    if (hiTemp.containsDigits()) "$hiTemp°" else WeatherIcons.PLACEHOLDER,
                    if (loTemp.containsDigits()) "$loTemp°" else WeatherIcons.PLACEHOLDER
                )
            )
        }

        if (feelsLikeModel != null) {
            if (conditionTempBuilder.isNotEmpty()) {
                conditionTempBuilder.append(" ${WeatherIcons.PLACEHOLDER} ")
            }

            conditionTempBuilder.append("${feelsLikeModel.label}: ${feelsLikeModel.value}")
        }

        if (conditionTempBuilder.isNotEmpty()) {
            updateViews.setTextViewText(R.id.condition_temp, conditionTempBuilder)
        } else {
            updateViews.setViewVisibility(R.id.condition_temp, View.GONE)
        }

        if (chanceModel != null) {
            updateViews.setImageViewResource(
                R.id.weather_popicon,
                wip.getWeatherIconResource(chanceModel.icon)
            )
            updateViews.setTextViewText(R.id.weather_pop, chanceModel.value)
            updateViews.setViewVisibility(R.id.weather_pop_layout, View.VISIBLE)
        } else {
            updateViews.setViewVisibility(R.id.weather_pop_layout, View.GONE)
        }

        // Extras
        val bigUpdateViews = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            RemoteViews(updateViews)
        } else {
            updateViews.clone()
        }

        if (windModel != null) {
            val windIconResId = wip.getWeatherIconResource(WeatherIcons.WIND_DIRECTION)
            if (windModel.iconRotation != 0) {
                bigUpdateViews.setImageViewBitmap(
                    R.id.weather_windicon,
                    rotateBitmap(
                        bitmapFromDrawable(context, windIconResId),
                        windModel.iconRotation.toFloat()
                    )
                )
            } else {
                bigUpdateViews.setImageViewResource(R.id.weather_windicon, windIconResId)
            }
            var speed = if (TextUtils.isEmpty(windModel.value)) "" else windModel.value.toString()
            speed = speed.split(",".toRegex()).first()
            bigUpdateViews.setTextViewText(R.id.weather_windspeed, speed)
            bigUpdateViews.setViewVisibility(R.id.weather_wind_layout, View.VISIBLE)
        } else {
            bigUpdateViews.setViewVisibility(R.id.weather_wind_layout, View.GONE)
        }

        if (humidityModel != null) {
            bigUpdateViews.setImageViewResource(
                R.id.humidity_icon,
                wip.getWeatherIconResource(humidityModel.icon)
            )
            bigUpdateViews.setTextViewText(R.id.humidity, humidityModel.value)
            bigUpdateViews.setViewVisibility(R.id.humidity_layout, View.VISIBLE)
        }

        if (windGustModel != null) {
            bigUpdateViews.setImageViewResource(
                R.id.windgust_icon,
                wip.getWeatherIconResource(windGustModel.icon)
            )
            bigUpdateViews.setTextViewText(R.id.windgust, windGustModel.value)
            bigUpdateViews.setViewVisibility(R.id.windgust_layout, View.VISIBLE)
        }

        bigUpdateViews.setViewVisibility(
            R.id.extra_layout,
            if (windModel != null || humidityModel != null || windGustModel != null) View.VISIBLE else View.GONE
        )

        val smallIconResId = when (settingsManager.getNotificationIcon()) {
            SettingsManager.TEMPERATURE_ICON -> {
                val tempLevel = temp.replace("°", "").toIntOrNull()
                if (tempLevel == null) {
                    R.drawable.notification_temp_unknown
                } else {
                    WeatherNotificationTemp.getTempDrawable(tempLevel)
                }
            }
            else /* SettingsManager.CONDITION_ICON */ -> {
                if (wim.isFontIcon) {
                    weatherIconResId
                } else {
                    // Use default icon pack here; animated icons are not supported here
                    wip.getWeatherIconResource(viewModel.weatherIcon)
                }
            }
        }

        val isTempIcon = settingsManager.getNotificationIcon() == SettingsManager.TEMPERATURE_ICON

        val contentIntent: PendingIntent = run {
            val onClickIntent = Intent(context, MainActivity::class.java)
                .setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
            var flags = 0
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S)
                flags = flags or PendingIntent.FLAG_IMMUTABLE
            PendingIntent.getActivity(context, 0, onClickIntent, flags)
        }

        // Builds the notification and issues it.
        val notifColor = if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
            val tempFloat = temp.toFloatOrNull()
            if (tempFloat != null) {
                val tempF = if (viewModel.curTemp!!.endsWith(Units.FAHRENHEIT)) {
                    tempFloat
                } else {
                    ConversionMethods.CtoF(tempFloat)
                }
                getColorFromTempF(tempF)
            } else {
                Colors.SIMPLEBLUE
            }
        } else {
            0 // Don't colorize; use Material You colors
        }

        val forecastType = settingsManager.getNotificationForecastType()
        val forecastItemLayoutId = if (wim.isFontIcon) {
            R.layout.weather_notification_forecast_item_tintable
        } else {
            R.layout.weather_notification_forecast_item
        }

        when (forecastType) {
            SettingsManager.NOTIF_FORECAST_NONE -> {
                bigUpdateViews.setViewVisibility(R.id.forecast_layout, View.GONE)
            }
            SettingsManager.NOTIF_FORECAST_DAILY -> {
                val forecasts = getForecasts(location, MAX_FORECASTS - 1)

                bigUpdateViews.removeAllViews(R.id.forecast_container)
                forecasts.forEach {
                    addForecastItem(
                        bigUpdateViews,
                        R.id.forecast_container,
                        forecastItemLayoutId,
                        it
                    )
                }

                bigUpdateViews.setViewVisibility(
                    R.id.forecast_layout,
                    if (forecasts.isNotEmpty()) View.VISIBLE else View.GONE
                )
            }
            SettingsManager.NOTIF_FORECAST_HOURLY -> {
                val hourlyForecasts = getHourlyForecasts(location, MAX_FORECASTS)

                bigUpdateViews.removeAllViews(R.id.forecast_container)
                hourlyForecasts.forEach {
                    addForecastItem(
                        bigUpdateViews,
                        R.id.forecast_container,
                        forecastItemLayoutId,
                        it
                    )
                }

                bigUpdateViews.setViewVisibility(
                    R.id.forecast_layout,
                    if (hourlyForecasts.isNotEmpty()) View.VISIBLE else View.GONE
                )
            }
        }

        return if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
            NotificationCompat.Builder(context, notChannelID)
                .setCustomContentView(updateViews)
                .setCustomBigContentView(bigUpdateViews)
                .setPriority(NotificationCompat.PRIORITY_LOW)
                .setOngoing(true)
                .setShowWhen(false)
                .setSmallIcon(smallIconResId)
                .setColor(notifColor)
                .setContentIntent(contentIntent)
                .build()
        } else {
            // Android 12 limits customizable notification area
            val contentTitle = if (isTempIcon) {
                viewModel.location
            } else {
                String.format(
                    "%s° - %s",
                    if (!temp.isNullOrWhitespace()) temp else WeatherIcons.PLACEHOLDER,
                    viewModel.location
                )
            }
            val contentText = if (feelsLikeModel != null) {
                "$condition ${WeatherIcons.PLACEHOLDER} ${feelsLikeModel.label} ${feelsLikeModel.value}"
            } else {
                condition
            }

            Notification.Builder(context, notChannelID)
                .setCustomContentView(null)
                .setContentTitle(contentTitle)
                .setContentText(contentText)
                .setCustomBigContentView(bigUpdateViews)
                .setOngoing(true)
                .setShowWhen(true)
                .setSmallIcon(smallIconResId)
                .setLargeIcon(Icon.createWithResource(context, weatherIconResId))
                .setColor(notifColor)
                .setContentIntent(contentIntent)
                .build()
        }
    }

    private fun addForecastItem(
        remoteViews: RemoteViews,
        @IdRes containerId: Int,
        @LayoutRes layoutId: Int,
        forecast: BaseForecastItemViewModel
    ) {
        val forecastItem = RemoteViews(remoteViews.`package`, layoutId)

        forecastItem.setTextViewText(R.id.forecast_date, forecast.shortDate)
        forecastItem.setTextViewText(R.id.forecast_hi, forecast.hiTemp)
        if (forecast is ForecastItemViewModel) {
            forecastItem.setTextViewText(R.id.forecast_lo, forecast.loTemp)
        }

        // WeatherIcon
        val wim = sharedDeps.weatherIconsManager
        val weatherIconResId = wim.getWeatherIconResource(forecast.weatherIcon)
        forecastItem.setImageViewResource(R.id.forecast_icon, weatherIconResId)

        if (forecast is HourlyForecastItemViewModel) {
            forecastItem.setViewVisibility(R.id.forecast_divider, View.GONE)
            forecastItem.setViewVisibility(R.id.forecast_lo, View.GONE)
        }

        remoteViews.addView(containerId, forecastItem)
    }

    private suspend fun getForecasts(
        locData: LocationData,
        forecastLength: Int
    ): List<ForecastItemViewModel> {
        val fcastData = settingsManager.getWeatherForecastData(locData.query)
        val fcasts = fcastData?.forecast

        if (fcasts?.isNotEmpty() == true) {
            return ArrayList<ForecastItemViewModel>(forecastLength).also {
                for (i in 0 until min(forecastLength, fcasts.size)) {
                    it.add(ForecastItemViewModel(fcasts[i]))
                }
            }
        }

        return emptyList()
    }

    private suspend fun getHourlyForecasts(
        locData: LocationData,
        forecastLength: Int
    ): List<HourlyForecastItemViewModel> {
        val now = ZonedDateTime.now().withZoneSameInstant(locData.tzOffset)
        val hrInterval = weatherModule.weatherManager.getHourlyForecastInterval()
        val hrfcasts = settingsManager.getHourlyForecastsByQueryOrderByDateByLimitFilterByDate(
            locData.query,
            forecastLength,
            now.minusHours((hrInterval * 0.5).toLong()).truncatedTo(ChronoUnit.HOURS)
        )

        if (hrfcasts.isNotEmpty()) {
            return ArrayList<HourlyForecastItemViewModel>(forecastLength).apply {
                var count = 0
                for (fcast in hrfcasts) {
                    add(HourlyForecastItemViewModel(fcast))
                    count++

                    if (count >= forecastLength) break
                }
            }
        }

        return emptyList()
    }
}